src/nunavut/lang/cpp/support/serialization.hpp

```diff
@@ -1,73 +1,318 @@
+    const std::uint32_t  round_mask = ~0xFFFU;
+
+    internal::Fp32 in;
+    NativeType     out;
+
+    in.f = value;
+
+    std::uint32_t sign = in.u & sign_mask;
+    in.u ^= sign;
+
+    if (in.u >= f32infty.u) /* Inf or NaN (all exponent bits set) */
+    {
+        /* NaN->sNaN and Inf->Inf */
+        out = (in.u > f32infty.u) ? 0x7FFFU : 0x7C00U;
+    }
+    else /* (De)normalized number or zero */
+    {
+        in.u &= round_mask;
+        in.f *= magic.f;
+        in.u -= round_mask;
+        if (in.u > f16infty.u)
         {
-            // we don't need to reevaluate the while condition.
-            break;
+            in.u = f16infty.u; /* Clamp to signed infinity if overflowed */
         }
-    } while (true);
 
-    return bits_copied;
+        out = 0xFFFF & static_cast<NativeType>(in.u >> 13); /* Take the bits! */
+    }
+
+    out = 0xFFFF & static_cast<NativeType>(out | (sign >> 16));
+
+    return out;
+}
+
+template <typename NativeType>
+float halfToNativeIeee(NativeType value)
+{
+    static_assert(sizeof(NativeType) >= 2, "NativeType must be at least 16 bits.");
+    static_assert(std::numeric_limits<NativeType>::is_integer, "NativeType must be integral.");
+    static_assert(!std::numeric_limits<NativeType>::is_signed, "NativeType must be unsigned.");
+
+    /*
+     * https://gist.github.com/rygorous/2144712
+     * Public domain, by Fabian "ryg" Giesen
+     */
+    const internal::Fp32 magic      = {(254U - 15U) << 23};
+    const internal::Fp32 was_infnan = {(127U + 16U) << 23};
+    internal::Fp32       out;
+
+    out.u = (value & 0x7FFFU) << 13; /* exponent/mantissa bits */
+    out.f *= magic.f;                /* exponent adjust */
+    if (out.f >= was_infnan.f)       /* make sure Inf/NaN survive */
+    {
+        out.u |= 255U << 23;
+    }
+    out.u |= (value & 0x8000U) << 16; /* sign bit */
+
+    return out.f;
+}
+
+template <unsigned BitLen>
+static typename NativeFloatSelector<BitLen>::int_storage_type toIeee(
+    typename NativeFloatSelector<BitLen>::float_type value)
+{
+    internal::enforceIeee<BitLen>();
+    union
+    {
+        typename NativeFloatSelector<BitLen>::int_storage_type i;
+        typename NativeFloatSelector<BitLen>::float_type       f;
+    } u;
+    static_assert(sizeof(u.f) * 8 == BitLen, "Float type selected does not match BitLen");
+    u.f = value;
+    return u.i;
 }
 
+template <unsigned BitLen>
+static typename NativeFloatSelector<BitLen>::float_type toNative(
+    typename NativeFloatSelector<BitLen>::int_storage_type value)
+{
+    internal::enforceIeee<BitLen>();
+    union
+    {
+        typename NativeFloatSelector<BitLen>::int_storage_type i;
+        typename NativeFloatSelector<BitLen>::Type             f;
+    } u;
+    static_assert(sizeof(u.f) * 8 == BitLen, "Float type selected does not match BitLen");
+    u.i = value;
+    return u.f;
+}
+
+}  // namespace IEEE754Converter
+
+template <typename ByteType, std::size_t static_capacity_bits>
```

Comment:
Why do we need to parametrize the byte type?

Reply:
Going to delete all this for now.